independent-projects/bootstrap/runner/src/main/java/io/quarkus/bootstrap/runner/JarFileReference.java

```diff
@@ -0,0 +1,72 @@
+package io.quarkus.bootstrap.runner;
+
+import java.io.IOException;
+import java.nio.file.Path;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.jar.JarFile;
+
+import io.smallrye.common.io.jar.JarFiles;
+
+public class JarFileReference {
+    // Guarded by an atomic reader counter that emulate the behaviour of a read/write lock.
+    // To enable virtual threads compatibility and avoid pinning it is not possible to use an explicit read/write lock
+    // because the jarFile access may happen inside a native call (for example triggered by the RunnerClassLoader)
+    // and then it is necessary to avoid blocking on it.
+    private final JarFile jarFile;
+
+    private final AtomicInteger readersCounter = new AtomicInteger(0);
```

Comment:
This obviously makes sense, but I don't think it will make much of a difference since you are not likely to have many more than 100 jar files.

Reply:
I agree with @geoand that this will introduce further complexity (and also the need of accessing the field through reflection) for a very limited memory gain, but I'm open to change this if required.

Reply:
The same trick also applies here, that you can use `AtomicInteger()` to make it slightly faster